### PR TITLE
Switch scheduler to use generated listers/informers

### PR DESCRIPTION
**Pull request by ncdc**
Where possible, switch the scheduler to use generated listers and
informers. There are still some places where it probably makes more
sense to use one-off reflectors/informers (listing/watching just a
single node, listing/watching scheduled & unscheduled pods using a field
selector).

I think this can wait until master is open for 1.7 pulls, given that we're close to the 1.6 freeze.

After this and #41482 go in, the only code left that references legacylisters will be federation, and 1 bit in a stateful set unit test (which I'll clean up in a follow-up).

@resouer I imagine this will conflict with your equivalence class work, so one of us will be doing some rebasing 😄 

cc @wojtek-t @gmarek  @timothysc @jayunit100 @smarterclayton @deads2k @liggitt @sttts @derekwaynecarr @kubernetes/sig-scheduling-pr-reviews @kubernetes/sig-scalability-pr-reviews

### CODE REVIEW

**Comment by jayunit100** [comment]
Thanks for this, happy to help, not sure what the review criteria are b/c im a little fuzzy on when/how generated components should be used...  just ping if anything we need to look at or help with though .

**Comment by ncdc** [comment]
@jayunit100 if we have generated components, like we now do for listers and shared informers, we should be motivated to convert to using them so we can remove the older hand-written listers/informers.

Also see https://github.com/kubernetes/kubernetes/pull/36238#discussion_r101284414 and https://github.com/kubernetes/kubernetes/pull/36238#issuecomment-280040996 for suggested guidelines on when to use informers. They work nicely if you want to react to a good portion of a data set. If, on the other hand, you only want to list/watch & react to changes to a single item, or a small set of items, a shared informer might not be the best fit if the data set itself is significantly larger.

**Comment by smarterclayton** [comment]
This is such a performance win I'd be inclined to get it in 1.6

**Comment by ncdc** [comment]
> This is such a performance win I'd be inclined to get it in 1.6

@smarterclayton I don't think this PR actually does all that much to help performance. The scheduler is currently running a bunch of reflectors/informers, typically 1 per type. It doesn't really have 5 reflectors all watching nodes, for example. All I'm really doing here is trying to get rid of the `legacylisters` package, as well as the hand-written informers.

But we'll see how it does (once I fix the startup issues).

**Comment by smarterclayton** [comment]
It avoids a fair amount of copies with the change to the pointer semantics, which should reduce CPU usage in high throughput scenarios.

**Comment by ncdc** [comment]
Good to know. Also, more fun with magic invisible api installation, ugh.

**Comment by ncdc** [comment]
@k8s-bot gci gce e2e test this

**Comment by ncdc** [comment]
@k8s-bot kubemark e2e test this
@k8s-bot gci gke e2e test this
@k8s-bot cvm gke e2e test this
@k8s-bot verify test this
@k8s-bot cvm gce e2e test this
@k8s-bot gce etcd3 e2e test this

**Comment by ncdc** [comment]
@k8s-bot unit test this
@k8s-bot verify test this
@k8s-bot gce etcd3 e2e test this
@k8s-bot gci gke e2e test this
@k8s-bot kubemark e2e test this
@k8s-bot cvm gke e2e test this

**Comment by ncdc** [comment]
@k8s-bot node e2e test this
@k8s-bot kops aws e2e test this
@k8s-bot non-cri node e2e test this
@k8s-bot non-cri e2e test this

**Comment by ncdc** [comment]
Working on fixing the vet issues now

**Comment by ncdc** [comment]
@k8s-bot bazel test this

```
--- FAIL: TestFindOldReplicaSets (0.00s)
	deployment_util_test.go:605: In test case "Get old ReplicaSets with two new ReplicaSets, only the oldest new ReplicaSet is seen as new ReplicaSet", expected []*v1beta1.ReplicaSet{(*v1beta1.ReplicaSet)(0xc42037d500), (*v1beta1.ReplicaSet)(0xc42037ce00)}, got []*v1beta1.ReplicaSet{(*v1beta1.ReplicaSet)(0xc42037ce00), (*v1beta1.ReplicaSet)(0xc42037d500)}: <nil>
FAIL
```

**Comment by ncdc** [comment]
@smarterclayton all tests are green. I'd really like someone from @kubernetes/sig-scheduling-pr-reviews to review this too if possible.

File: plugin/pkg/scheduler/algorithm/predicates/predicates.go [hunk 1]
```diff
@@ -58,13 +59,22 @@ type PersistentVolumeInfo interface {
 	GetPersistentVolumeInfo(pvID string) (*v1.PersistentVolume, error)
 }
 
+// CachedPersistentVolumeInfo implements PersistentVolumeInfo
+type CachedPersistentVolumeInfo struct {
+	corelisters.PersistentVolumeLister
+}
+
+func (c *CachedPersistentVolumeInfo) GetPersistentVolumeInfo(pvID string) (*v1.PersistentVolume, error) {
```
Comment (deads2k) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: The ID here is actually a name?

Comment (deads2k) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: I'm assuming you're conforming to some interface here?

Comment (ncdc) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: Yes, ID is a name.

See comment on line 62 re interface.

**Comment by deads2k** [comment]
They were logically running a cache before, it was just slightly less specific.

/lgtm
/approve

**Comment by resouer** [comment]
Changes lgtm, my rebase is on the way 📆

**Comment by davidopp** [comment]
cc/ @bsalamat This might be related to the thing you were asking about today.

**Comment by davidopp** [comment]
cc/ @wojtek-t

**Pull request merged**